include/seastar/net/api.hh

```diff
@@ -243,6 +243,11 @@ public:
     explicit operator bool() const noexcept {
         return static_cast<bool>(_csi);
     }
+    /// Waits for the peer of this \c connected_socket to close
+    ///
+    /// \return future that resolves when the peer closes its end or shuts down
+    /// the socket for writing (see man poll about POLLRDHUP for details)
+    future<> wait_closed();
```

Comment:
Isn't this duplicated with reading from the local input_stream (that represents the remote write end) until we reach eof?

What happens if there's some unconsumed data in the local input_stream buffer and the remote shuts down its writing end? We'll get RDHUP but it is premature, since we didn't consume all the data yet.

Reply:
> Isn't this duplicated with reading from the local input_stream (that represents the remote write end) until we reach eof?

It is. Consider http client case -- caller created a socket and made a successful request, then want to keep socket alive until next request. Before 2nd request is made server closes the connection, so client has two options:
1. keep socket open until 2nd request is made. In this case client-side socket holds resource until 2nd request which can take time
2. keep background reading fiber that notices EOF and disposes the socket. In this case receive path will render extra latencies like send path did in #1067

> What happens if there's some unconsumed data in the local input_stream buffer and the remote shuts down its writing end? We'll get RDHUP but it is premature, since we didn't consume all the data yet.

Yes, FIN is reported instantly, not queued. But why is that a problem?

```
    while (true) {
        client = pool.get();
        resp = client.send_request();
        body = resp.in().read_all();
        pool.put(client);
    }
    ...
    client pool::get() {
        if (list_of_clients.empty()) {
            client = new client();
            client.wait_closed([&client] {
                client.dead = true;
                if (!client.active) {
                    client.close();
                }
            });
        } else {
            client = list_of_clients.pop_front();
        }
        client.active = true;
        return client
    }
    ...
    void pool::put(client) {
        if (!client.dead) {
            client.active = false;
            list_of_clients.push_back(client);
        } else {
            client.close();
        }
    }
```

Reply:
> > Isn't this duplicated with reading from the local input_stream (that represents the remote write end) until we reach eof?
> 
> It is. Consider http client case -- caller created a socket and made a successful request, then want to keep socket alive until next request. Before 2nd request is made server closes the connection, so client has two options:
> 
> 1. keep socket open until 2nd request is made. In this case client-side socket holds resource until 2nd request which can take time
> 2. keep background reading fiber that notices EOF and disposes the socket. In this case receive path will render extra latencies like send path did in [RPC send path is prone to producing extra latency #1067](https://github.com/scylladb/seastar/issues/1067)

Agree both are bad.

> 
> > What happens if there's some unconsumed data in the local input_stream buffer and the remote shuts down its writing end? We'll get RDHUP but it is premature, since we didn't consume all the data yet.
> 
> Yes, FIN is reported instantly, not queued. But why is that a problem?
> 
> ```
>     while (true) {
>         client = pool.get();
>         resp = client.send_request();
>         body = resp.in().read_all();
>         pool.put(client);
>     }
>     ...
>     client pool::get() {
>         if (list_of_clients.empty()) {
>             client = new client();
>             client.wait_closed().then([&client] {
>                 client.dead = true;
>                 if (!client.active) {
>                     client.close();
>                 }
>             });
>         } else {
>             client = list_of_clients.pop_front();
>         }
>         client.active = true;
>         return client
>     }
>     ...
>     void pool::put(client) {
>         if (!client.dead) {
>             client.active = false;
>             list_of_clients.push_back(client);
>         } else {
>             client.close();
>         }
>     }
> ```

Didn't follow the code closely, but I understand the user is supposed to perform a logical when_all() on the reader fiber and on the wait_closed() future.

After the clarification, I agree with the patch.

Please update the documentation to note the subtle point that when_closed() does not mean there is no more data, just that reading from the input_stream (or data_source) is expected to now complete until eof without waiting for the network.

Reply:
Also, only a single wait_closed() can be outstanding, but it can run in parallel with regular reads/writes.

Should we name it more clearly? wait_peer_closed()? wait_peer_closed_write_end()?

Reply:
> Also, only a single wait_closed() can be outstanding, but it can run in parallel with regular reads/writes.

Yes.

> Should we name it more clearly? wait_peer_closed()? wait_peer_closed_write_end()?

wait_input_shutdown() maybe? It also wakes up if connected_socket::shutdown_input() is called.